onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+    node->mode = cmodes[i];
+    node->is_not_leave = node->mode != NODE_MODE::LEAF;
+    node->truenode = NULL;   // nodes_truenodeids[i];
+    node->falsenode = NULL;  // nodes_falsenodeids[i];
+    node->missing_tracks = i < static_cast<size_t>(nodes_missing_value_tracks_true.size())
+                               ? (nodes_missing_value_tracks_true[i] == 1
+                                      ? MissingTrack::TRUE
+                                      : MissingTrack::FALSE)
+                               : MissingTrack::NONE;
+    node->is_missing_track_true = node->missing_tracks == MissingTrack::TRUE;
+    if (idi.find(node->id) != idi.end()) {
+      ORT_THROW("Node ", node->id.node_id, " in tree ", node->id.tree_id, " is already there.");
+    }
+    idi.insert(std::pair<TreeNodeElementId, TreeNodeElement<OTYPE>*>(node->id, node));
+  }
+
+  TreeNodeElementId coor;
+  TreeNodeElement<OTYPE>* it;
+  for (i = 0; i < static_cast<size_t>(n_nodes_); ++i) {
+    it = nodes_ + i;
+    if (!it->is_not_leave)
+      continue;
+    coor.tree_id = it->id.tree_id;
+    coor.node_id = static_cast<int>(nodes_truenodeids[i]);
+
+    auto found = idi.find(coor);
+    if (found == idi.end()) {
+      ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (truenode).");
+    }
+    if (coor.node_id >= 0 && coor.node_id < n_nodes_) {
+      it->truenode = found->second;
+      if ((it->truenode->id.tree_id != it->id.tree_id) ||
+          (it->truenode->id.node_id == it->id.node_id)) {
+        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
+      }
+    } else
+      it->truenode = NULL;
+
+    coor.node_id = static_cast<int>(nodes_falsenodeids[i]);
+    found = idi.find(coor);
+    if (found == idi.end()) {
+      ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (falsenode).");
+    }
+    if (coor.node_id >= 0 && coor.node_id < n_nodes_) {
+      it->falsenode = found->second;
+      if ((it->falsenode->id.tree_id != it->id.tree_id) ||
+          (it->falsenode->id.node_id == it->id.node_id)) {
+        ORT_THROW("One falsenode is pointing either to itself, either to another tree.");
+      }
+    } else
+      it->falsenode = NULL;
+  }
+
+  int64_t previous = -1;
+  for (i = 0; i < static_cast<size_t>(n_nodes_); ++i) {
+    if ((previous == -1) || (previous != nodes_[i].id.tree_id))
+      roots_.push_back(nodes_ + i);
+    previous = nodes_[i].id.tree_id;
+  }
+
+  TreeNodeElementId ind;
+  SparseValue<OTYPE> w;
+  for (i = 0; i < target_class_nodeids.size(); i++) {
+    ind.tree_id = static_cast<int>(target_class_treeids[i]);
+    ind.node_id = static_cast<int>(target_class_nodeids[i]);
+    if (idi.find(ind) == idi.end()) {
+      ORT_THROW("Unable to find node ", coor.tree_id, "-", coor.node_id, " (weights).");
+    }
+    w.i = target_class_ids[i];
+    w.value = target_class_weights[i];
+    idi[ind]->weights.push_back(w);
+  }
```

Comment:
Does the push_back involve a copy when compiled in release mode? 
That could maybe be avoided if this was idi[ind]->weights.push_back({target_class_ids[i], target_class_weights[i]})
May not be significant per-wise though

Reply:
Not sure about that. idi[ind]->weights will contain less than n_targets, but most of the time only one.